esiclient/v1/cluster/cluster.py

```diff
@@ -13,18 +13,70 @@
 import concurrent.futures
 import json
 import logging
-import time
 
 from osc_lib.command import command
 from osc_lib.i18n import _
 
+from oslo_utils import uuidutils
+
 from esiclient import utils
 
 
+ESI_CLUSTER_UUID = 'esi_cluster_uuid'
+ESI_TRUNK_UUID = 'esi_trunk_uuid'
+ESI_PORT_UUID = 'esi_port_uuid'
+ESI_FIP_UUID = 'esi_fip_uuid'
+
+
 class ESIOrchestrationException(Exception):
     pass
 
 
+class List(command.Lister):
+    """List ESI clusters"""
+
+    log = logging.getLogger(__name__ + ".List")
+
+    def get_parser(self, prog_name):
+        parser = super(List, self).get_parser(prog_name)
+        return parser
+
+    def take_action(self, parsed_args):
+        self.log.debug("take_action(%s)", parsed_args)
+
+        ironic_client = self.app.client_manager.baremetal
+
+        nodes = ironic_client.node.list(fields=['uuid', 'name', 'extra'])
+
+        cluster_dict = {}
+        for node in nodes:
+            extra = node.extra
+            # only look at nodes that specify a cluster UUID
+            if ESI_CLUSTER_UUID in extra:
+                cluster_uuid = extra[ESI_CLUSTER_UUID]
+                esi_extra = {}
+                for key in extra:
+                    # only display 'extra' attributes set by ESI
+                    if key.startswith('esi') and key != ESI_CLUSTER_UUID:
+                        esi_extra[key] = extra[key]
+                if cluster_uuid not in cluster_dict:
+                    cluster_dict[cluster_uuid] = {}
+                cluster_dict[cluster_uuid][node.name] = esi_extra
+
+        cluster_info = []
+        for cluster_uuid in cluster_dict:
+            node_info = []
+            extra_info = []
+            for node in cluster_dict[cluster_uuid]:
+                node_info.append(node)
+                extra_info.append(str(cluster_dict[cluster_uuid][node]))
+            cluster_info.append([cluster_uuid,
+                                 "\n".join(node_info),
+                                 "\n".join(extra_info)])
+
+        return ["Cluster", "Node", "Associated"], cluster_info
```

Comment:
The current `associated` info contains all nodes' info within a cluster. Do we want to list each node and its associated data together?

Reply:
Oh, the way the `node_info` and `extra_info` is joined, it works pretty much as you say: each node's associated resources are printed on the same line